[Bugfix](load) fix streamload failure due to false unhealthy replica in concurrent stream load

yixiutt opened the pull request:
in concurrent stream load, fe will run publish version task concurrently,
which cause publish task disorder in be.
For example:
fe publish task with version 1 2 3 4
be may handle task with sequence 1 2 4 3
In case above, when report tablet info, be found that version 4
published but version 3 not visable, it'll report version miss to fe,
and fe will set replica lastFailedVersion, and finally makes transaction
commits fail while no quorum health replicas。

Add a strategy to handle this case, be must publish version in sequence,
if publish task version not match tablet max_version, it will be push
back to queue and wait.

# Proposed changes

Issue Number: close #xxx

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

dataroaring commented on be/src/agent/task_worker_pool.cpp:
```diff
@@ -673,38 +673,50 @@ void TaskWorkerPool::_publish_version_worker_thread_callback() {
+        std::vector<TTabletId> error_tablet_ids;
+        Status res = Status::OK();
+        while (!current_finished) {
+            {
```
Could we remove this while?

zhannngchen commented on be/src/agent/task_worker_pool.cpp:
```diff
@@ -673,38 +673,50 @@ void TaskWorkerPool::_publish_version_worker_thread_callback() {
+                    current_finished = true;
+                    break;
+                } else if (res.precise_code() == OLAP_ERR_ROWSET_VERSION_NOT_CONTINUOUS) {
+                    std::unique_lock<std::mutex> worker_thread_lock(_worker_thread_lock);
```
Add comments here？

zhannngchen commented on be/src/agent/task_worker_pool.cpp:
```diff
@@ -673,38 +673,50 @@ void TaskWorkerPool::_publish_version_worker_thread_callback() {
+                    break;
+                } else if (res.precise_code() == OLAP_ERR_ROWSET_VERSION_NOT_CONTINUOUS) {
+                    std::unique_lock<std::mutex> worker_thread_lock(_worker_thread_lock);
+                    _tasks.push_back(agent_task_req);
```
we should also add a log here.

dataroaring commented on be/src/agent/task_worker_pool.cpp:
```diff
@@ -707,6 +708,14 @@ void TaskWorkerPool::_publish_version_worker_thread_callback() {
+            // push back to task queue if publish version not continuous
+            // make sure publish version task run in order
+            _tasks.push_back(agent_task_req);
+            _worker_thread_condition_variable.notify_one();
```
This thread is working, so we donot need notify here.

yixiutt replied:
fixed

yixiutt commented on be/src/olap/task/engine_publish_version_task.cpp:
```diff
@@ -83,6 +83,12 @@ Status EnginePublishVersionTask::finish() {
+            {
+                std::shared_lock rdlock(tablet->get_header_lock());
+                if (tablet->max_version().second + 1 != par_ver_info.version) {
+                    return Status::OLAPInternalError(OLAP_ERR_ROWSET_VERSION_NOT_CONTINUOUS);
```
we should retrun OLAPInternalError here so add a new error type

dataroaring reviewed (dismissed):
LGTM

yiguolei commented on be/src/agent/task_worker_pool.cpp:
```diff
@@ -707,6 +708,13 @@ void TaskWorkerPool::_publish_version_worker_thread_callback() {
+            std::unique_lock<std::mutex> worker_thread_lock(_worker_thread_lock);
+            // push back to task queue if publish version not continuous
+            // make sure publish version task run in order
+            _tasks.push_back(agent_task_req);
```
Maybe there will be a dead loop. If current tablet is a bad tablet(it really has a missed version), you will push the task to task queue and not publish it. Clone process will failed.

yiguolei reviewed (changes requested):
waiting

Commit:
[bugfix](load) fix streamload failure due to false unhealthy replica in concurrent stream load

in concurrent stream load, fe will run publish version task concurrently,
which cause publish task disorder in be.
For example:
fe publish task with version 1 2 3 4
be may handle task with sequence 1 2 4 3
In case above, when report tablet info, be found that version 4
published but version 3 not visable, it'll report version miss to fe,
and fe will set replica lastFailedVersion, and finally makes transaction
commits fail while no quorum health replicas。

Add a time condition if a version miss for 60 seconds then report version miss.

morningman commented on be/src/olap/tablet.cpp:
```diff
@@ -1277,7 +1279,10 @@ bool Tablet::_contains_rowset(const RowsetId rowset_id) {
+// need check if consective version missing in full report
+// alter tablet will ignore this check
+void Tablet::build_tablet_report_info(TTabletInfo* tablet_info,
+                                      bool enable_consective_missing_check) {
```
```suggestion
                                      bool enable_consecutive_missing_check) {
```

morningman commented on be/src/olap/tablet.cpp:
```diff
@@ -1290,7 +1295,23 @@ void Tablet::build_tablet_report_info(TTabletInfo* tablet_info) {
+                _last_missed_version = cversion.second + 1;
+                _last_missed_time_s = MonotonicSeconds();
+            }
+        } else {
```
This `else` block should be with `if (enable_consective_missing_check) {`

morningman replied:
And please add comment here to explain why we need this logic

morningman reviewed (approved):
LGTM

yiguolei reviewed (approved):
LGTM
